### PR TITLE
1955 temperature sensor graphql for logs and breaches

### PR DESCRIPTION
**Pull request by adrianwb**
Part of #1955

# 👩🏻‍💻 What does this PR do? 
 
Added graphql/service for temperature log, breach and  breach config

# 🧪 How has/should this change been tested? 

Just cargo test

## 💌 Any notes for the reviewer?

Not finished yet, but creating the PR now so that others can work on it if I'm picked as a juror on Monday and unavailable...

Still todo:
- fix failing graphql tests (doesn't work if I try to add timestamps for permission tests)
- add filter by type for breach and breach config
- add filter by location_id for breach and log
- update unique check for breach to include type as well as sensor_id, start/end timestamp and duration

Temperature logs should only be inserted via a sensor update (reading the USB data), and should have a sensor_id, timestamp and temperature. Its store_id and location_id are inherited from the sensor. If the sensor location is changed, the location of the temperature log should also be changed, but only if it is blank. No user-updatable fields.

Temperature breach configs should only be inserted via a sensor update (reading the USB data), and should have a type, duration and max/min temperature. Its store_id is inherited from the sensor. Only its description and active status should be updatable by an end user with the right permissions.

Temperature breaches should only be inserted via a sensor update (reading the USB data), and should have a sensor_id, start/end timestamp and duration and type. Its store_id and location_id are inherited from the sensor. If the sensor location is changed, the location of the temperature breach should also be changed, but only if it is blank. Its threshold max/min and duration are inherited from the breach config used to generate it. Only its acknowledgement status should be updatable by an end user with the right permissions

## 📃 Documentation

_No user facing changes_

or

- [ ] Functional change 1
- [ ] Functional change 2

### CODE REVIEW

**Comment by andreievg** [comment]
> Only its acknowledgement status should be updatable by an end user with the right permissions

Do fridgetags provide acknowledgement status ?

We are planning to leave out omSupply breach acknowledgement from MVP. (only device acknowledgements would be displayed, i.e. ones that were acknowledged in mSupply Cold Chain)

**Comment by andreievg** [comment]
Thanks @adrianwb for doing all this work, I started reviewing it and very quickly realised that I am repeating myself a lot and thought it's best summarised in overall comment and maybe a chat.

So the main point for me, is we want to be really targeted with PRs and code that is added to the code base, this helps with PR review (we have very limited PR review capacity atm), stale code, compilation and just generally being efficient. I did mention this in previous PR, sorry I didn't manage to emphasise this point enough and maybe failed to summarise MVP requirements in the epic (although most of the areas where this comment applies are not listed in the epic).

Basically we only need API for this:

Query for logs

Query for temperature breaches

Query for sensors

Update sensor (location, is active and description/name)

For the above 3 we only need the following loaders:

Log -> Sensor
Sensor -> Location
Log -> Breach 
Breach -> Sensor

We would need filters for those (basically logs and breaches would need to be filterable by sensor, location and breach types). Btw we can even do nested repository queries, [something like this](https://github.com/openmsupply/open-msupply/blob/56c2e7abb7d8d1ad0b9d1212168218e7834a703b/server/repository/src/db_diesel/program_requisition/program_requisition_settings.rs#L28)

There is not need for any other mutations or queries, although I see you've gone to some length to add mutations, I think we would only use the temperature breach config at some point in the future, but I think the others would need to be removed. Sorry I missed this bit, i did mention it a couple of times [here](https://github.com/openmsupply/open-msupply/issues/1798#issuecomment-1659591411) and [here](https://github.com/openmsupply/open-msupply/issues/1955), maybe that's not the right medium to communicate this requirements or perhaps i was to vague ? The idea is the logs and breaches are introduced directly with the server, server would get the file uploaded and the integrate logs, breaches, sensors and configs.

We can make a note of the commit that added things that are removed from this PR that maybe useful in the future (like the breach config mutations), I think a comment in the epic might be sufficient.

I was wondering how to move forward with this PR, we are starting on front end work for Cold Chain and would need the queries and sensor mutation. Is it ok for us or you to remove the other bits that are not used, so we can start from something that's easier to digest ?

**Comment by adrianwb** [comment]
> > Only its acknowledgement status should be updatable by an end user with the right permissions
> 
> Do fridgetags provide acknowledgement status ?

I haven't tried it on a live fridgetag but AFAIK, they do

> We are planning to leave out omSupply breach acknowledgement from MVP. (only device acknowledgements would be displayed, i.e. ones that were acknowledged in mSupply Cold Chain)

Yes, that's fine -> no need for any breach mutations then

**Comment by adrianwb** [comment]
> Thanks @adrianwb for doing all this work, I started reviewing it and very quickly realised that I am repeating myself a lot and thought it's best summarised in overall comment and maybe a chat.
> 
> So the main point for me, is we want to be really targeted with PRs and code that is added to the code base, this helps with PR review (we have very limited PR review capacity atm), stale code, compilation and just generally being efficient. I did mention this in previous PR, sorry I didn't manage to emphasise this point enough and maybe failed to summarise MVP requirements in the epic (although most of the areas where this comment applies are not listed in the epic).
> 
> Basically we only need API for this:
> 
> Query for logs
> 
> Query for temperature breaches
> 
> Query for sensors
> 
> Update sensor (location, is active and description/name)
> 
> For the above 3 we only need the following loaders:
> 
> Log -> Sensor; Sensor -> Location; Log -> Breach; Breach -> Sensor

I'm not entirely sure what you mean by "loaders" - can you give an example?

> We would need filters for those (basically logs and breaches would need to be filterable by sensor, location and breach types). Btw we can even do nested repository queries, [something like this](https://github.com/openmsupply/open-msupply/blob/56c2e7abb7d8d1ad0b9d1212168218e7834a703b/server/repository/src/db_diesel/program_requisition/program_requisition_settings.rs#L28)

That's interesting. Don't forget filtering by timestamp(s) as well - that was one of things I was having trouble figuring out...

> There is not need for any other mutations or queries, although I see you've gone to some length to add mutations, I think we would only use the temperature breach config at some point in the future, but I think the others would need to be removed. Sorry I missed this bit, i did mention it a couple of times [here](https://github.com/openmsupply/open-msupply/issues/1798#issuecomment-1659591411) and [here](https://github.com/openmsupply/open-msupply/issues/1955), maybe that's not the right medium to communicate this requirements or perhaps i was to vague ? The idea is the logs and breaches are introduced directly with the server, server would get the file uploaded and the integrate logs, breaches, sensors and configs.

Most likely my bad for not fully understanding what I was doing! There aren't any log mutations and I didn't spend too much effort on the breach/config mutations anyway - mostly a copy and edit job from similar instances - so no problem to take them out again.

> We can make a note of the commit that added things that are removed from this PR that maybe useful in the future (like the breach config mutations), I think a comment in the epic might be sufficient.

Sounds fine

> I was wondering how to move forward with this PR, we are starting on front end work for Cold Chain and would need the queries and sensor mutation. Is it ok for us or you to remove the other bits that are not used, so we can start from something that's easier to digest ?

As I understand it, @roxy-dao has taken this on already (#2275 is branched off this) - happy for her (or anyone else) to remove the bits that aren't needed at this stage and to fix up the filters as needed. I'm concentrating now on integrating the Berlinger crate into omSupply, so that it can be called via the sensor service => that will do the sensor read and upsert the four sensor tables accordingly. Changing the sensor location should probably be a separate issue, but I'd imagine that it would also be done via the sensor service?

**Comment by roxy-dao** [comment]
Removed most the mutations I think. Will make the other changes in #2275

File: server/graphql/temperature_breach_config/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,211 @@
+use async_graphql::*;
```
Comment (andreievg) [inline server/graphql/temperature_breach_config/src/lib.rs:1]: @roxy-dao technically don't need now either

File: server/graphql/types/src/types/permissions.rs [hunk 2]
```diff
@@ -20,6 +20,10 @@ pub enum UserPermission {
     LocationMutate,
     SensorMutate,
     SensorQuery,
+    TemperatureBreachConfigMutate,
```
Comment (andreievg) [inline server/graphql/types/src/types/permissions.rs:23]: This one is not needed and the config query

File: server/repository/src/db_diesel/user_permission_row.rs [hunk 3]
```diff
@@ -34,8 +34,10 @@ pub enum Permission {
     SensorQuery,
     // no existing mSupply permissions for temperature log/breach (which shouldn't be user editable)
     // but one for temperature_breach_config,
-    //TemperatureBreachConfigMutate,
-    //TemperatureBreachConfigQuery,
+    TemperatureBreachConfigMutate,
```
Comment (andreievg) [inline server/repository/src/db_diesel/user_permission_row.rs:37]: Config query and mutate not needed atm

File: server/service/src/auth.rs [hunk 4]
```diff
@@ -37,6 +37,14 @@ pub enum Resource {
     // sensor
     QuerySensor,
     MutateSensor,
+    // temperature log
+    QueryTemperatureLog,
+    // temperature breach
+    QueryTemperatureBreach,
+    MutateTemperatureBreach,
+    // temperature breach config
+    QueryTemperatureBreachConfig,
```
Comment (andreievg) [inline server/service/src/auth.rs:46]: This one too I think, can be left out for now

**Review by andreievg — APPROVED** [review]

**Review by andreievg — COMMENTED** [review]
Thanks, looks good.

I've made a comment about config permissions and config query, i think we can remove that one, it's ok to do it in a new issue to avoid the need to propograte all the changes.

There is also question about the loaders being use (or not), will look at that in the downstream PR

**Review by andreievg — APPROVED** [review]

**Review by andreievg — APPROVED** [review]

**Pull request merged**